### PR TITLE
Modify acl hook behavior

### PR DESCRIPTION
**Pull request by WinterNis**

### CODE REVIEW

File: hook/acl_test.go [hunk 1]
```diff
@@ -100,108 +87,74 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
-	netns.Set(s.randomNamespace)                         // Enter random namespace
-	rules, _ := s.iptablesDriver.List("filter", "INPUT") // Retrieve chain rules
-	assert.Equal(s.T(), []string{
-		"-P INPUT ACCEPT",
-		"-A INPUT -s 10.0.0.1/32 -j ACCEPT",
-	}, rules) // Ensure rules have been well inserted
+	assert.Subset(s.T(),
+		append(
+			baseInputRule,
+			[]string{
```
Comment (janisz) [inline hook/acl_test.go]: Do you need to wrap this in slice if you use `...`?

File: hook/acl.go [hunk 2]
```diff
@@ -34,32 +36,40 @@ var ACLHook = Hook{
...
-			if label.GetKey() != aclHookLabel {
+			if match == nil {
 				continue
 			}
 
-			// Stop here if label is set but null
-			if label.GetValue() == "" {
-				break
+			// Check that port index is valid and match port mapping
+			var port mesos.ContainerInfo_DockerInfo_PortMapping
```
Comment (janisz) [inline hook/acl.go]: `port` → `portMapping` the you read GetHostPort from port, it looks weird.

File: hook/acl.go [hunk 3]
```diff
@@ -34,32 +36,40 @@ var ACLHook = Hook{
...
+			var portIndex int
+			portIndex, err = strconv.Atoi(match[1])
```
Comment (tgermain) [inline hook/acl.go]: check `match` len before accessing it.

**Review by tgermain — CHANGES_REQUESTED** [review]

File: hook/acl.go [hunk 4]
```diff
@@ -12,14 +13,15 @@ import (
...
-	"github.com/vishvananda/netns"
 	"go.uber.org/zap"
 )
 
 const (
-	aclHookLabel = "EXECUTOR_ALLOWED_IP"
+	aclHookRuleTemplate = "-i %s -p %s -s %s --dport %s -j ACCEPT"
 )
 
+var aclHookLabel = regexp.MustCompile("ACL_(?P<portIndex>[0-9]+)_ALLOWED")
```
Comment (Devatoria) [inline hook/acl.go]: EXECUTOR_port_ACL

File: hook/acl.go [hunk 5]
```diff
@@ -34,129 +36,171 @@ var ACLHook = Hook{
...
 }
 
-// injectRuleIntoNamespace injects an iptables rule into the given namespace
-//TODO: use namespace package instead of manually entering namespace
-func injectRuleIntoNamespace(ns netns.NsHandle, rule string) error {
-	// Get current namespace
-	currentNs, err := netns.Get()
-	if err != nil {
-		return err
+func checkChain(driver *iptables.IPTables) (string, error) {
```
Comment (Devatoria) [inline hook/acl.go]: Please comment this function

File: hook/acl.go [hunk 6]
```diff
@@ -34,129 +36,171 @@ var ACLHook = Hook{
...
+	}
 
-	// Lock OS thread to avoid accidental namespace switching
-	runtime.LockOSThread()
-	defer runtime.UnlockOSThread()
+	for i := range chains {
+		if chains[i] == aclChain {
+			return aclChain, nil
+		}
+	}
```
Comment (Devatoria) [inline hook/acl.go]: Please jump line

File: hook/acl.go [hunk 7]
```diff
@@ -34,129 +36,171 @@ var ACLHook = Hook{
...
+	return "", fmt.Errorf("Chain %s does not exists", aclChain)
+}
 
-	// Switch on given namespace
-	err = netns.Set(ns)
-	if err != nil {
-		return err
+func generateACL(
```
Comment (Devatoria) [inline hook/acl.go]: Please comment function

File: hook/acl.go [hunk 8]
```diff
@@ -34,129 +36,171 @@ var ACLHook = Hook{
...
 
-	// Inject rule
-	ruleParts := strings.Split(rule, " ")
-	driver, err := iptables.New()
-	if err != nil {
-		return err
+	// Get task container ports
+	portMappings := info.GetContainer().GetDocker().GetPortMappings()
+
+	for _, label := range info.GetLabels().GetLabels() {
```
Comment (Devatoria) [inline hook/acl.go]: Please comment this (very long) loop

File: hook/acl.go [hunk 9]
```diff
@@ -34,129 +36,171 @@ var ACLHook = Hook{
...
+				)
+				err = action("filter", chain, strings.Split(aclRule, " ")...)
+				if err != nil {
+					if stopOnError {
+						return fmt.Errorf("Error while injecting acl iptables rule: %v", err)
+					}
+				}
+			}
+		}
+	}
```
Comment (Devatoria) [inline hook/acl.go]: Please jump line

File: hook/acl_test.go [hunk 10]
```diff
@@ -15,83 +15,88 @@ import (
...
+		case "acl.chain":
+			return s.aclChain
+		default:
+			return viper.GetString(arg)
+		}
+	})
+
 	driver, _ := iptables.New() // Get iptables driver
 	s.iptablesDriver = driver
+
```
Comment (Devatoria) [inline hook/acl_test.go]: Empty line

File: hook/acl_test.go [hunk 11]
```diff
@@ -100,108 +105,141 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
-		"-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT", // Default related/established traffic allow rule
-		"-A INPUT -j DROP",                                        // Default DROP rule
-	}, rules)
+	assert.Subset(s.T(),
+		append(
+			baseInputRule,
+			"-A INPUT -s 8.8.8.8/32 -i ext1 -p tcp -m tcp --dport 10000 -j ACCEPT",
+			"-A INPUT -s 10.0.0.0/24 -i ext1 -p tcp -m tcp --dport 10000 -j ACCEPT",
+		),
+		rules)
```
Comment (Devatoria) [inline hook/acl_test.go]: Jump a line between rules and )

File: hook/acl_test.go [hunk 12]
```diff
@@ -100,108 +105,141 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
+	assert.Error(s.T(), s.hook.RunPostRun(s.c, info, ""))
+	rules, _ = s.iptablesDriver.List("filter", "FORWARD")
+	assert.Subset(s.T(), baseForwardRule, rules)
+
+	// Error is thrown if chain is output
+	s.aclChain = "OUTPUT"
+	assert.Error(s.T(), s.hook.RunPostRun(s.c, info, ""))
+	rules, _ = s.iptablesDriver.List("filter", "OUTPUT")
+	assert.Subset(s.T(), baseOutputRule, rules)
+
```
Comment (Devatoria) [inline hook/acl_test.go]: Empty line

File: hook/acl_test.go [hunk 13]
```diff
@@ -100,108 +105,141 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
-		"-A INPUT -j DROP",                                           // Default DROP rule
-	}, rules)
-	netns.Set(s.isolatedNamespace)
+	assert.Subset(s.T(),
+		append(
+			baseInputRule,
+			"-A INPUT -s 8.8.8.8/32 -i all -p tcp -m tcp --dport 10000 -j ACCEPT",
+			"-A INPUT -s 10.0.0.0/24 -i all -p tcp -m tcp --dport 10000 -j ACCEPT",
+		),
+		rules)
```
Comment (Devatoria) [inline hook/acl_test.go]: LJ between rules and )

File: hook/acl_test.go [hunk 14]
```diff
@@ -100,108 +105,141 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
-	rules, _ := s.iptablesDriver.List("filter", "INPUT") // Retrieve chain rules
-	assert.Equal(s.T(), []string{
-		"-P INPUT ACCEPT",
-		"-A INPUT -s 10.0.0.1/32 -j ACCEPT",
-	}, rules) // Ensure rules have been well inserted
+// Check that :
+// - hook does nothing when container is in host network
+// - acls are correctly removed at task hook execution
+func (s *ACLHookTestSuite) TestACLHookRunPreStop() {
+
```
Comment (Devatoria) [inline hook/acl_test.go]: Empty line

File: hook/acl_test.go [hunk 15]
```diff
@@ -100,108 +105,141 @@ func (s *AclHookTestSuite) TestACLHookExecute() {
...
+	// Execute remove acl hook to remove injected iptables
+	assert.Nil(s.T(), s.hook.RunPreStop(s.c, info, ""))
+	inputRules, _ := s.iptablesDriver.List("filter", "INPUT")
+
+	assert.Equal(
+		s.T(),
+		referenceInputRule,
+		inputRules,
+	)
+
```
Comment (Devatoria) [inline hook/acl_test.go]: Empty line

**Review by Devatoria — CHANGES_REQUESTED** [review]

**Review by Devatoria — APPROVED** [review]

**Pull request merged**